File: docs/code/stack.py
class Stack:

    def __init__(self, data=None):
        self.data = []
        if data:
            for i in data:
                self.push(i)
        
    def push(self, v):
        self.data.append(v)
    
    def pop(self):
        return self.data.pop()
    
    def __len__(self):
        return len(self.data)
    
    def __iter__(self):
        while self:
            yield self.pop()

File: code/c/stack/stack.c
#include "stack.h"

#include <assert.h>
#include <stdlib.h>

#define T cii_stack_t
struct T {
    int count; // 比用size要妥当一些。
    struct node_s {
        void *data;
        struct node_s *next;
    } *head;
};

T stack_new(void)
{
    T stack = malloc(sizeof(*stack));
    assert(stack);
    stack->head = NULL;
    stack->count = 0;
    return stack;
}

bool stack_empty(T stack)
{
    assert(stack);
    return stack->count == 0;
}

void stack_push(T stack, void *x)
{
    struct node_s *t;

    assert(stack);
    t = malloc(sizeof(*t));
    assert(t);
    t->data = x;
    t->next = stack->head;
    stack->head = t;
    stack->count++;
}

void *stack_pop(T stack)
{
    void *x;
    struct node_s *t;

    assert(stack);
    assert(stack->count > 0);
    t = stack->head;
    stack->head = t->next;
    stack->count--;
    x = t->data;
    free(t);
    return x;
}
